go/cmd/vtbackup/vtbackup.go

```diff
@@ -14,120 +14,331 @@ See the License for the specific language governing permissions and
+			return fmt.Errorf("can't initialize database with reparent journal: %v", err)
 		}
+		// Now we're ready to take the backup.
+		name := backupName(time.Now(), tabletAlias)
+		if err := mysqlctl.Backup(ctx, mycnf, mysqld, logutil.NewConsoleLogger(), backupDir, name, *concurrency, extraEnv); err != nil {
+			return fmt.Errorf("backup failed: %v", err)
+		}
+		log.Info("Initial backup successful.")
+		return nil
+	}
 
+	// Restore from backup.
+	dbName := *initDbNameOverride
+	if dbName == "" {
+		dbName = fmt.Sprintf("vt_%s", *initKeyspace)
+	}
+
+	log.Infof("Restoring latest backup from directory %v", backupDir)
+	restorePos, err := mysqlctl.Restore(ctx, mycnf, mysqld, backupDir, *concurrency, extraEnv, map[string]string{}, logutil.NewConsoleLogger(), true, dbName)
+	switch err {
+	case nil:
+		log.Infof("Successfully restored from backup at replication position %v", restorePos)
 	case mysqlctl.ErrNoBackup:
-		// No-op, starting with empty database.
+		return fmt.Errorf("no backup found; not starting up empty since -initial_backup flag was not enabled")
 	case mysqlctl.ErrExistingDB:
-		// No-op, assuming we've just restarted.  Note the
-		// replication reporter may restart replication at the
-		// next health check if it thinks it should. We do not
-		// alter replication here.
+		return fmt.Errorf("can't run vtbackup because data directory is not empty")
 	default:
-		log.Fatalf("Error restoring backup: %v", err)
+		return fmt.Errorf("can't restore from backup: %v", err)
+	}
+
+	// We have restored a backup. Now start replication.
+	if err := resetReplication(ctx, restorePos, mysqld); err != nil {
+		return fmt.Errorf("error resetting replication: %v", err)
+	}
+	if err := startReplication(ctx, mysqld, topoServer); err != nil {
+		return fmt.Errorf("error starting replication: %v", err)
+	}
+
+	// Get the current master replication position, and wait until we catch up
+	// to that point. We do this instead of looking at Seconds_Behind_Master
+	// (replication lag reported by SHOW SLAVE STATUS) because that value can
+	// sometimes lie and tell you there's 0 lag when actually replication is
+	// stopped. Also, if replication is making progress but is too slow to ever
+	// catch up to live changes, we'd rather take a backup of something rather
+	// than timing out.
+	tmc := tmclient.NewTabletManagerClient()
+	masterPos, err := getMasterPosition(ctx, tmc, topoServer)
+	if err != nil {
+		return fmt.Errorf("can't get the master replication position: %v", err)
 	}
 
-	// We have restored a backup ( If one existed, now make sure replicatoin is started
-	if err := startReplication(context.Background(), pos, mysqld, topoServer); err != nil {
-		log.Fatalf("Error Starting Replication %v", err)
+	// Remember the time when we fetched the master position, not when we caught
+	// up to it, so the timestamp on our backup is honest (assuming we make it
+	// to the goal position).
+	backupTime := time.Now()
+
+	if restorePos.Equal(masterPos) {
```

Comment:
Just to double check, this will never result in you removing your last backup as long as minRetentionCount is 1 ( it looks to be that way )

Reply:
Right. The pruning is orthogonal to whether you just took a new backup or not. It's not going to get tricked into thinking we just created a new backup, because it looks at the full list of backups right before making its decisions.